rules/opentitan.bzl

```diff
@@ -43,15 +45,166 @@ PER_DEVICE_DEPS = {
+            ],
+            PROD = [
+                create_prod_key("spx_fake_prod_key_0", "@//sw/device/silicon_creator/rom/keys/fake/spx:prod_key_0_spx"),
+            ],
+        ),
+        REAL = None,
+        UNAUTHORIZED = [
+            create_key_("spx_unauthorized_0", "@//sw/device/silicon_creator/rom/keys/spx:unauthorized_0_spx", []),
+        ],
+    ),
+)
+
+SILICON_OWNER_KEYS = struct(
+    RSA = struct(
+        FAKE = struct(
+            TEST = [
+                create_test_key("rsa_fake_rom_ext_test_key_0", "@//sw/device/silicon_creator/rom_ext/keys/fake:rom_ext_test_private_key_0"),
+            ],
+            DEV = [
+                create_dev_key("rsa_fake_rom_ext_dev_key_0", "@//sw/device/silicon_creator/rom_ext/keys/fake:rom_ext_dev_private_key_0"),
+            ],
+            PROD = None,
+        ),
+        REAL = None,
+        UNAUTHORIZED = None,
+    ),
+)
+
+def zip(l1, l2):
+    length = min(len(l1), len(l2))
+    return [(l1[i], l2[i]) for i in range(length)]
+
+def flatten(l):
+    return [item for ll in l for item in ll]
+
+def key_allowed_in_lc_state(key, hw_lc_state_val):
+    all_hw_lc_state_vals = structs.to_dict(CONST.LCV).values()
+    if not hw_lc_state_val in all_hw_lc_state_vals:
+        fail("Wrong life cycle state value: '{}', must be one of {}. Did you pass a string instead of the integer value?".format(hw_lc_state_val, all_hw_lc_state_vals))
+    return hw_lc_state_val in key.hw_lc_states
+
+def get_keysets_for_lc_state(hw_lc_state, rsa = SILICON_CREATOR_KEYS.RSA.FAKE, spx = SILICON_CREATOR_KEYS.SPX.FAKE):
+    # [(rsa_key, spx_key), ...]
+    if rsa and spx:
+        rsa = flatten(structs.to_dict(s = rsa).values())
+        spx = flatten(structs.to_dict(s = spx).values())
+    elif rsa and not spx:
+        rsa = flatten(structs.to_dict(s = rsa).values())
+        spx = [None] * len(rsa)
+    elif not rsa and spx:
+        spx = flatten(structs.to_dict(s = spx).values())
+        rsa = [None] * len(spx)
+    else:
+        fail("No keys were provided")
+
+    if len(rsa) != len(spx):
+        fail("Lists have different lengths")
+    keys = zip(rsa, spx)
+    res = [
+        create_keyset(rsa, spx)
+        for rsa, spx in keys
+        if (rsa == None or key_allowed_in_lc_state(rsa, hw_lc_state)) and (spx == None or key_allowed_in_lc_state(spx, hw_lc_state))
+    ]
+    return res
+
+def filter_keysets_for_lc_state(keysets, hw_lc_state):
+    return [k for k in keysets if (not k.rsa or key_allowed_in_lc_state(k.rsa, hw_lc_state)) and (not k.spx or key_allowed_in_lc_state(k.spx, hw_lc_state))]
+
+# Default keys used to sign BL0 images for testing.
+RSA_ONLY_KEYSETS = [
+    create_keyset(SILICON_CREATOR_KEYS.RSA.FAKE.TEST[0], None),
+    create_keyset(SILICON_CREATOR_KEYS.RSA.FAKE.DEV[0], None),
+    create_keyset(SILICON_CREATOR_KEYS.RSA.FAKE.PROD[0], None),
+    create_keyset(SILICON_CREATOR_KEYS.RSA.UNAUTHORIZED[0], None),
+]
```

Comment:
Not sure if Bazel/Skylark supports this between `.bzl` and `BUILD` files, but I think it would be nice if this was a named dictionary -- that way in the `BUILD` file, it would be `RSA_ONLY_KEYSETS["FAKE_TEST"]` instead of `RSA_ONLY_KEYSETS[0]`.

Reply:
Good point. I think it does since we use dicts in `BUILD` files but this list exists to preserve the inputs of the existing e2e tests.